### PR TITLE
Add smart weight initialization

### LINKED ISSUE #6: Starting weight initialization
Look for a better way of initializing neural network weights.

### PR DESCRIPTION
**Pull request by szymanskir**
Achieved a score of 0.9566 for the regular configuration of 784, 30, 10 and 100 epochs. An improvement compared to the initial strategy of randomizing weights (the previous score was 0.9238).

Resolves #6

### CODE REVIEW

File: backpropagation/network/neural_network.py [hunk 1]
```diff
@@ -41,16 +41,34 @@ def __init__(
         self.layers_count = len(neurons_count_per_layer)
 
         self.weights = [
-            np.random.normal(size=(
+            self._create_weight_matrix(size=(
                 neurons_count_per_layer[layer + 1],
-                neurons_count_per_layer[layer] + 1
+                neurons_count_per_layer[layer]
```
Comment (rzepinskip) [inline backpropagation/network/neural_network.py]: Why the difference?

File: backpropagation/network/neural_network.py [hunk 2]
```diff
@@ -41,16 +41,34 @@ def __init__(
...
+            Args:
+                size (Tuple[int, int]): Size of the layers between which the matrix
+                should be initialized.
+
+            Returns (np.array): Weight matrix used to calculate the activation from one
+                layer to another.
+        """
+        rows, columns = size
+        biases = np.random.normal(size=(rows, 1))
+        weight_matrix = np.random.normal(size=(rows, columns), scale=1/np.sqrt(columns))
```
Comment (rzepinskip) [inline backpropagation/network/neural_network.py]: What is the reason behind this specific scale parameter?

File: backpropagation/network/neural_network.py [hunk 3]
```diff
@@ -41,16 +41,34 @@ def __init__(
...
             for layer in range(self.layers_count - 1)
         ]
-
         self.activation_function = activation_function
         self.cost_function = cost_function
 
+    def _create_weight_matrix(self, size: Tuple[int, int]) -> np.array:
+        """Creates a weight matrix between layers of sizes specified by ``size``.
+
+            It creates the biases weights and stacks it horizontally in front of the
```
Comment (rzepinskip) [inline backpropagation/network/neural_network.py]: If `biases = np.random.normal(size=(rows, 1))` it seems like it's column vector, which cannot not be added horizontally.

File: backpropagation/network/neural_network.py (hunk 2 above)
Comment (szymanskir) [reply backpropagation/network/neural_network.py]: You make the distribution "flatter" and by that neurons are less likely to saturate. You can read more on that here: http://neuralnetworksanddeeplearning.com/chap3.html#weight_initialization

File: backpropagation/network/neural_network.py (hunk 3 above)
Comment (szymanskir) [reply backpropagation/network/neural_network.py]: In numpy horizontal stacking refers to situation where a, b are stacked like this: `a|b`. I wanted to remain consistent with the code.

File: backpropagation/network/neural_network.py (hunk 1 above)
Comment (szymanskir) [reply backpropagation/network/neural_network.py]: Now there is a separate function for creating the weight matrices which automatically adds the bias terms. Previously the random.normal function was used from the numpy library hence it had to be written explicitly that we need an extra column for bias terms.

**Pull request merged**